player/js/utils/text/SkriptShaper.js

```diff
@@ -0,0 +1,999 @@
+    this.layoutPerformed = false;
+    this.properties = [];
+    this.graphemes = [];
+    this.words = [];
+    this.runs = [];
+    this.lines = [];
+    this.bounds = new Rect(0, 0, 0, 0);
+  }
+
+  /**
+   * Adds styled text part to the paragraph
+   * @param {String} text Text to add
+   * @param {String} fontStyle Style of the text
+   */
+  addText(text, fontStyle) {
+    if (this.layoutPerformed) {
+      alert('Cannot modify the text after layout');
+      return;
+    }
+    const textUnit = new TextRange(this.text.length, this.text.length + text.length);
+    this.fontStyleRanges.push(new FontStyleRanges(textUnit, fontStyle));
+    this.text += text;
+  }
+
+  /**
+   * @param {GlyphLine} line
+   * @param {Number} end
+   */
+  finishLine(line, end) {
+    if (line.textRange.start === end) {
+      // The line has not started really
+      return;
+    }
+    // Make sure the grapheme, the run and the word are finished by now
+    line.glyphRange.end = this.graphemes.length;
+    line.runRange.end = this.runs.length;
+    line.wordRange.end = this.words.length;
+    this.lines.push(structuredClone(line));
+  }
+
+  /**
+   * @param {GlyphRun} run
+   * @param {Number} end
+   */
+  finishRun(run, end) {
+    if (run.textRange.start === end) {
+      // The run has not really started
+      return;
+    }
+    // Make sure the grapheme and the word are finished by now
+    run.textRange.end = end;
+    run.glyphRange.end = this.graphemes.length;
+    run.wordRange.end = this.words.length;
+    this.runs.push(structuredClone(run));
+  }
+
+  /**
+   * @param {GlyphWord} word
+   */
+  finishWord(word) {
+    // Make sure the grapheme is finished
+    word.glyphRange.end = this.graphemes.length;
+    this.words.push(structuredClone(word));
+  }
+
+  /**
+   * @param {GlyphCluster} cluster
+   * @param {Number} textStart
+   * @param {Number} textEnd
+   * @param {Number} graphemeStart
+   * @param {Number} graphemeEnd
+   * @param {String} text
+   * @param {Boolean} whitespaces
+   * @param {Boolean} newLines
+   */
+  addCluster(cluster, textStart, textEnd, graphemeStart, graphemeEnd, text, whitespaces, newLines) {
+    cluster.textRange = { start: textStart, end: textEnd };
+    cluster.glyphRange = { start: graphemeStart, end: graphemeEnd };
+    cluster.text = text;
+    cluster.isWhitespaces = whitespaces;
+    cluster.isNewLine = newLines;
+    this.graphemes.push(structuredClone(cluster));
+  }
+
+  /**
+   * Performs the text layout
+   * @param {Number} width
+   */
+  layout(width, fontManager) {
+    // TODO: deal with locales
+    this.clearOutputs();
+    this.extractSegments('grapheme', Properties.graphemeStart);
+    this.extractSegments('word', Properties.wordStart);
+
+    let span;
+    let rect = null;
+    if (this.coloredId !== undefined) {
+      span = document.getElementById(this.coloredId);
+      rect = document.getElementById(this.measurementId);
+    } else {
+      this.coloredId = 'undefined';
```

Comment:
are quotes around undefined intentional?

Reply:
This code should go away -- it's just a leftovers of an old way of testing.